src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs

```diff
@@ -0,0 +1,511 @@
+
+            // Set the caret position to the properly indented column in the desired line.
+            var newDocument = document.WithText(formattedText);
+            var newDocumentText = await newDocument.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var caretPosition = GetIndentedLinePosition(newDocument, newDocumentText, desiredCaretLine.LineNumber, cancellationToken);
+
+            // The new line edit is calculated against the original text, d0, to get text d1.
+            // The formatting edits are calculated against d1 to get text d2.
+            // Merge the formatting and new line edits into a set of whitespace only text edits that all apply to d0.
+            var overallChanges = newLineEdit != null ? MergeFormatChangesIntoNewLineChange(newLineEdit.Value, formattingChanges) : formattingChanges;
+            return new BraceCompletionResult(overallChanges, caretPosition);
+
+            static bool HasExistingEmptyLineBetweenBraces(int openingPoint, int closingPoint, SourceText sourceText)
+            {
+                // Check if there is an empty new line between the braces already.  If not insert one.
+                // This handles razor cases where they insert an additional empty line before calling brace completion.
+                var openingPointLine = sourceText.Lines.GetLineFromPosition(openingPoint).LineNumber;
+                var closingPointLine = sourceText.Lines.GetLineFromPosition(closingPoint).LineNumber;
+
+                return closingPointLine - 1 > openingPointLine && sourceText.Lines[closingPointLine - 1].IsEmptyOrWhitespace();
+            }
+
+            static TextLine GetLineBetweenCurlys(int closingPosition, SourceText text)
+            {
+                var closingBraceLineNumber = text.Lines.GetLineFromPosition(closingPosition - 1).LineNumber;
+                return text.Lines[closingBraceLineNumber - 1];
+            }
+
+            static LinePosition GetIndentedLinePosition(Document document, SourceText sourceText, int lineNumber, CancellationToken cancellationToken)
+            {
+                var indentationService = document.GetRequiredLanguageService<IIndentationService>();
+                var indentation = indentationService.GetIndentation(document, lineNumber, cancellationToken);
+
+                var baseLinePosition = sourceText.Lines.GetLinePosition(indentation.BasePosition);
+                var offsetOfBacePosition = baseLinePosition.Character;
+                var totalOffset = offsetOfBacePosition + indentation.Offset;
+                var indentedLinePosition = new LinePosition(lineNumber, totalOffset);
+                return indentedLinePosition;
+            }
+
+            static ImmutableArray<TextChange> MergeFormatChangesIntoNewLineChange(TextChange newLineEdit, ImmutableArray<TextChange> formattingEdits)
```

Comment:
i don't mind this.  but do we have appropriate testing of all the codepaths in this?

also, why was this better thna just exposing the existing merging function and just having us merge two lists togetehr where one list contains the single text change?

Reply:
Yeah I went this way for two main reasons
1.   Adding IVT from the compiler layer to the features layer here so I could use this caused a bunch errors across the sln due to types (like ArrayBuilder) that are part of both MS.CA.Workspaces and MS.CA.  When I have IVTs to both everything breaks `The type 'ArrayBuilder<T>' exists in both 'Microsoft.CodeAnalysis.Workspaces, Version=42.42.42.42, Culture=neutral, PublicKeyToken=31bf3856ad364e35' and 'Microsoft.CodeAnalysis, Version=42.42.42.42, Culture=neutral, PublicKeyToken=31bf3856ad364e35
`.  Not certain if there is a way to fix this, but making the whole type public didn't seem right.
2.  Based on your suggestion, this wasn't too complicated to write.

Will add tests

Reply:
my *strong* preference would be to extract the algorithm into a static class/helper that is then just linked in (like we do with ArrayBuilder itself).  I would not expect IVT, just a sharing of the code here.

Reply:
Ah OK, sounds good I'll do that.

Reply:
@CyrusNajmabadi do you know of a good existing project for shared items like this?  It looks like MS.CA is using
Microsoft.CodeAnalysis.PooledObjects.projitems
Microsoft.CodeAnalysis.Debugging.projitems
AnalyzerDriver.projitems

But none of those seemed quite correct, and I didn't see any common project that is shared.  I can create a new projectitem for this but wanted to check if I was missing one.

Reply:
I can look later. However, wherever the pile objects are shared seems like the right place